README.md
<div align="center">
<a href="https://github.com/eddiejibson/chae-limitrr"><img alt="chae" src="https://cdn.oxro.io/chae/img/limitrr.png" width="432.8" height="114.2"></a>
<br>
<br>
<img src="https://circleci.com/gh/eddiejibson/chae-limitrr.svg?style=svg"></img>
<img src="https://www.codefactor.io/repository/github/eddiejibson/chae-limitrr/badge">
<a href="https://paypal.me/eddiejibson/5"><img src="https://img.shields.io/badge/donate-PayPal-brightgreen.svg"></a>
<img src="https://requires.io/github/eddiejibson/chae-limitrr/requirements.svg?branch=master">
<img src="https://img.shields.io/npm/dw/limitrr.svg">
<p>Light Express rate limiting using Redis.</p>
</div>

Limitrr assists with the rate-limiting of various routes within your Express application. Unlike other similar packages, this utility allows the user to limit not only by the number of requests but also the number of completed actions (e.g allowing a certain amount of accounts to be successfully created within a timespan) and have such restricted with custom options. As well as this, custom discriminators are possible - you no longer have to limit by just the user's IP.

If you appreciate this project, please 🌟 it on GitHub.

**Pull Requests are welcomed**

## Roadmap

- [x]   Create multiple route limits (separate expiry times e.t.c) without having to initialize the Limitrr class multiple times
- [x]   Return headers to user with rate limiting details - how many requests/actions remain before restrictions are put in place, how long before the values expire and how many requests/actions are allowed per that route.
- [x]   Unit Tests
- [x]   Pass parameters into functions via an object

# Installation

``` bash
npm install limitrr --save
```

# Quick Guide

## Basic Usage

``` javascript
const express = require("express");
const app = express();
const Limitrr = require("limitrr");
const limitrr = new Limitrr({
    //Redis keystore connection information
    "redis": {
        "password": "supersecret",
        "host": "666.chae.sh",
        "port": 6379,
        "family": 4
        //There are many more options all of which can be seen further into the documentation under the "configuration" title
    },
    //General limitrr options
    "options": {
        "expiry": 900,
        //There are many more options all of which can be seen further into the documentation under the "configuration" title
    },
    "routes": {
        //Overwrite the default options
        "default": {
            //You do not have to overwrite all the options, only the
            //ones needed.
            "requestsPerExpiry": 500
        },
        "createUser": {
            //Not all keys need to be set here either, they will resolve
            //to the default if not set.
            "requestsPerExpiry": 300
        }
    }
});

app.use(limitrr.getIp()); //Middleware to get a user's IP - even if the application is behind Cloudflare
//This is assigned to req.realIp

//Simple rate limiting
app.get('/', limitrr.limit(), (req, res, next) => {
    res.send('Hello World!')
});

app.get("/registerUser/:user", limitrr.limit(), (req, res, next) => {
    //Non intensive actions like simple verification will have a different limit to intensive ones.
    //and will only be measured in terms of each request via the middleware.
    //No further action is required.
    if (req.params.user.length < 5) {
        //Dummy function creating user
        someRandomModule.registerUser().then((result) => {
            //Intensive actions like actually registering a user should have a
            //different limit to normal requests, hence the completedActionsPerExpiry option.
            //and should only be added to once this task has been completed fully
            //In this example, we will be limiting the amount of completed actions a certain IP can make.
            //Anything can be passed in here, however. For example, a email address or user ID.
            //req.realIp was determined by calling the middleware earlier - limitrr.getIp()
            limitrr.complete({
                "discriminator": req.realIp}
            }); //Calling will add to the completed count
            //Bearing in mind this a promise.
            limitrr.complete(req.realIp).then((result) => {
                //In this example, we will be returning a
                //success message as the action has been completed.
                return res.status(200).json({
                    "message": "Success!"
                });
            }).catch((err) => {
                //Handle error
            });
        });
    }
});

app.listen(3000, () => console.log(`Limitrr example app listening on port 3000!`))
```

## Headers

The limitrr.limit() middleware function will return headers to the user. They are defined as followed:

- `X-RateLimit-Limit`: What is the maximum amount of requests a user can make to this route before they are rate limited.
- `X-RateLimit-Remaining`: How many requests does the user have remaining before they are rate limited.
- `X-RateLimit-Reset`: How much time does the user have (in seconds) before their current count of requests are reset.
- `X-RateLimit-Limit-Actions`: What is the maximum amount of completed actions (e.g user registration) a user can complete via this route before they are rate limited.
- `X-RateLimit-Remaining-Actions`: How many completed actions can the user make before they are rate limited.
- `X-RateLimit-Reset-Actions`: How much time does the user have (in seconds) before their current count of completed actions are reset.
  
## Get the value of a certain key

### limitrr.get()

**Returns**: Promise

``` javascript
limitrr.get({
    "discriminator": discriminator, //Required
    "route": route, //Not required
    "type": type //Not required
});
```

#### Parameters

*Must be passed into function via object*

- **discriminator**: **Required** Where discriminator is the thing being limited (e.g x amount of completed actions/requests per discriminator)
- **route**: *String* When route should the values be retrieved from? If this is not set, it will get the counts from the default route.
- **type**: *String* Instead of retrieving both values, you can specify either `requests` or `completed` in this key and only that will be returned as an integer.

```javascript
limitrr.get({
    "discriminator": discriminator,
    "route": route
    "type": type
}); //Besides discriminator, all parameters are optional.
//If type is not passed into the function, it will
//return both the amount of requests and completed actions


//Where discriminator is the thing being limited
//e.g x amount of completed actions/requests per discriminator
limitrr.get({"discriminator": discriminator});

//This tends to be the user's IP.
limitrr.get({"discriminator": req.realIp})
//This will return both the amount of requests and completed actions stored under the
//discriminator provided in an object. You can handle like this:
limitrr.get({"discriminator": req.realIp}).then((res) => {
    console.log(`${res.requests} Requests`);
    console.log(`${res.completed} Completed Tasks`);
}).catch((err) => {
    //Handle error
});

//The above example would get the request and completed task count from the default
//route. If you would like to retrieve values from a different route, you can specify
//this as well. It can be done like this:
limitrr.get({
    "discriminator": req.realIp,
    "route": "exampleRouteName"
}).then((res) => {
    console.log(`${res.requests} Requests made through the route exampleRouteName`);
    console.log(`${res.completed} Completed Tasks made through the route exampleRouteName`);
}).catch((err) => {
    //Handle error
});

```

## Complete action/task count

### limitrr.complete()

**Returns**: Promise

```javascript
limitrr.complete({
    "discriminator": discriminator, //Required
    "route": route //Not Required
});
```

#### Parameters

*Must be passed into function via object*

- **discriminator**: **Required** Where discriminator is the thing being limited (e.g x amount of completed actions per discriminator)
- **route**: *String* What route should the values be inserted into? If this is not set, it will default to `default`.

## Removal of values from certain request/completed keys

### limitrr.reset()

**Returns**: Promise

``` javascript
limitrr.reset({
    "discriminator": discriminator, //Required
    "type": type, //Not required
    "route": route //Not required
});
```

#### Parameters

*Must be passed into function via object*

- **discriminator**: **Required** Where discriminator is the thing being limited (e.g x amount of completed actions/requests per discriminator)
- **type**: Which count do you wish to be reset? `requests` or `completed`? If this is not set, both will be removed.
- **route**: *String* What route should the values be retrieved from? If this is not set, it will get the counts from the `default` route.

``` javascript
//Where discriminator is the thing being limited
//e.g x amount of completed actions/requests per discriminator
//This will remove both the amount of requests and completed action count
limitrr.reset({"discriminator": discriminator});

//This tends to be the user's IP.
limitrr.reset({"discriminator": req.realIp});

//If you want to remove either one of the amount of requests or completed actions.
//but not the other, this can be done too.
//The value passed in can either be "requests" or "completed".
//In this example, we will be removing the request count for a certain IP
limitrr.reset({
    "discriminator": req.realIp,
    "type": "requests"
    }).then((res) => {
    if (res) {
        console.log("Requests removed");
    }
}).catch((err) => {
    //Handle error
});

//If you wish to reset counts from a particular route, this can be done as well.
limitrr.reset({
    "discriminator": req.realIp,
    "route": "exampleRouteName"
    }).then((res) => {
     if (res) {
        console.log("Requests removed from the route exampleRouteName");
    }
}).catch((err) => {
    //Handle error
});
```

# Configuration

## redis

**Required**: false

**Type**: Object

**Description**: Redis connection information.

- **port**: *Integer* Redis port. Defaults to: `6379`

- **host**: *String* Redis hostname. Defaults to: `"127.0.0.1"`

- **family**: *Integer* If Redis hostname is IPv4 (4) or IPv6 (6). Defaults to: `4` (IPv4)

- **password**: *String* Redis password. Defaults to: `""`

- **db**: *Integer* Redis DB. Defaults to: `0`

### Example of the redis object that could be passed into Limitrr

``` javascript
"redis": {
  "port": 6379, //Redis Port. Required: false. Defaults to 6379
  "host": "127.0.0.1", //Redis hostname. fequired: False. Defaults to "127.0.0.1".
  "family": 4, //Is redis hostname IPv4 (4) or IPv6 (6)? Required: false. Defaults to 4 (IPv4).
  "password": "mysecretpassword1234", //Redis password. Required: false. Defaults to "" (empty).
  "db": 0 //Redis DB. Required: false. Defaults to 0.
}
```

## options

**Required**: false

**Type**: Object

**Description**: Various options to do with Limitrr.

- **keyName**: *String* The keyname all of the requests will be stored under. This is mainly for aesthetic purposes and does not affect much. However, this should be changed on each initialization of the class to prevent conflict. Defaults to: `"limitrr"`
- **errorStatusCode**: *Integer* Status code to return when the user is being rate limited. Defaults to `429` (Too Many Requests)
- **catchErrors**: *Boolean* Should important errors such as failure to connect to the Redis keystore be caught and displayed? If this is set to false, it will throw an error instead. Defaults to `true`.

### Example of the options object that could be passed into Limitrr

``` javascript
"options": {
  "keyName": "myApp", //The keyname all of the requests will be stored under. Required: false. Defaults to "limitrr"
  "errorStatusCode": 429, //Status code to return when the user is being rate limited. Defaults to: 429 (Too many requests)
  "catchErrors": true //Should important errors such as failure to connect to the Redis keystore be caught and displayed?
}
```

## routes

**Required**: false

**Type**: Object

**Description**: Define route restrictions.

Inside the routes object, you can define many separate routes and set custom rules within them. The custom rules you can set are:

- **requestsPerExpiry**: *Integer* How many requests can be accepted until user is rate limited? Defaults to: `100`
- **completedActionsPerExpiry**: *Integer* How many completed actions can be accepted until the user is rate limited? This is useful for certain actions such as registering a user - they can have a certain amount of requests but a different (obviously smaller) amount of "completed actions". So if users have recently been successfully registered multiple times under the same IP (or other discriminator), they can be rate limited. They may be allowed 100 requests per certain expiry for general validation and the like, but only a small fraction of that for intensive procedures. Defaults to the value in `requestsPerExpiry` or `5` if not set.
- **expiry**: *Integer* How long should the requests be stored (in seconds) before they are set back to 0? If set to -1, values will never expire and will stay that way indefinitely or must be manually removed. Defaults to: `900` (15 minutes)
- **completedExpiry**: *Integer* How long should the "completed actions" (such as the amount of users registered from a particular IP or other discriminator) be stored for (in seconds) before it is set back to 0? If set to -1, such values will never expire and will stay that way indefinitely or must be manually removed. Defaults to the value in `expiry` or `900` (15 minutes) if not set.
- **errorMsgs**: *Object* Seperate error messages for too many requests and too many completed actions. They have been given the respective key names "requests" and "actions". This will be returned to the user when they are being rate limited. If no string was set in `requests`, it will default to `"As you have made too many requests, you are being rate limited."`. Furthermore, if a value has not been set in `completed`, it will resolve to the string found in `requests`. Or, if that wasn't set either, `"As you performed too many successful actions, you have been rate limited."` will be it's value.

## Example of the routes object

``` javascript
"routes": {
    //Overwrite default route rules - not all of the keys must be set,
    //only the ones you wish to overwrite
    "default": {
        "expiry": 1000
    },
    "exampleRoute": {
        "requestsPerExpiry": 100,
        "completedActionsPerExpiry": 5,
        "expiry": 900,
        "completedExpiry": 900,
        "errorMsgs": {
            "requests": "As you have made too many requests, you are being rate limited.",
            "completed": "As you performed too many successful actions, you have been rate limited."
        }
    },
    //If not all keys are set, they will revert to
    //the default values
    "exampleRoute2": {
        "requestsPerExpiry": 500
    }
}
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.0.0] - 2018-11-17
### Added
- Headers. Limitrr will now return various values such as the requests remaining for the user.
- Unit tests.
- Changelog.

### Changed
- Default method of passing options into a function is now via an object.


test.js
/*
 * @Project: limitrr-example
 * @Created Date: Sunday, November 4th 2018, 8:19:05 pm
 * @Author: Edward Jibson
 * @Last Modified Time: November 14th 2018, 4:54:58 pm
 * @Last Modified By: Edward Jibson
 * @Copyright: (c) 2018 Oxro Holdings LLC
 */
const express = require('express');
const app = express();
const Limitrr = require("./lib/limitrr");
const limitrr = new Limitrr({
    //Redis keystore connection information
    "redis": {
        "password": "AZnYJ04zdPqn7lex1MCP30Hj0dvw0ydUbTPFkSeCu6fJYoj18c14ANi9MhXct5jr86XWOi2cdayBxy8wOQug83JLosdxwenVp2bpd03RwhKcWzva3WJTrMgJhLBT6bfD",
        "host": "159.69.148.73",
        "port": 6379,
        "family": 4
        //There are many more options all of which can be seen further into the documentation under the "configuration" title
    },
    //General limitrr options
    "options": {
        "expiry": 900,
        //There are many more options all of which can be seen further into the documentation under the "configuration" title
    },
    "routes": {
        //Overwrite the default options
        "default": {
            //You do not have to overwrite all the options, only the
            //ones needed.
            "requestsPerExpiry": 500
        },
        "createUser": {
            //Not all keys need to be set here either, they will resolve
            //to the default if not set.
            "requestsPerExpiry": 300
        }
    }
});

app.use(limitrr.getIp()); //Sets req.realIp to user IP - helpful later on.

//Simple rate limiting
//If route is not set
app.get('/', limitrr.limit(), (req, res, next) => {
    res.send('Hello World!')
});

//We have specified the route in the limit function this time.
app.get("/registerUser/:user", limitrr.limit("createUser"), (req, res, next) => {
    //Non intensive actions like simple verification will have a different limit 
    //and will only be measured in terms of each request via the middleware
    if (req.params.user.length < 5) {
        //Dummy function creating user
        someRandomModule.registerUser().then((result) => {
            //Intensive actions like actually registering a user should have a different limit, hence the completedActionsPerExpiry option
            //and should only be added to once such task has been completed fully
            //In this example, we will be limiting the amount of completed actions a certain IP can make. 
            //Anything can be passed in here. For example, a email address or user ID.
            //req.realIp was determined by calling the middleware earlier - limitrr.getIp()
            limitrr.complete(req.realIp); //Calling such will complete
            //Is also a promise.
            limitrr.complete(req.realIp).then((result) => {
                return res.status(200).json({ //In this example, we will be returning a success message as the action has been completed.
                    "message": "Success!"
                });
            }).catch((err) => {
                //Do something with the caught error
            });
        });
    }
});

app.listen(8089, () => console.log(`Limitrr example app listening on port 8089!`))

lib/limitrr.js
/*
 * @Project: limitrr
 * @Created Date: Wednesday, October 31st 2018, 2:42:59 pm
 * @Author: Edward Jibson
 * @Last Modified Time: November 17th 2018, 8:03:32 pm
 * @Last Modified By: Edward Jibson
 */
const Redis = require("ioredis");

class limitrr {
    constructor(conf) {
        conf.routes = conf.routes || {};
        conf.routes.default = conf.routes.default || {};
        conf.routes.default.errorMsgs = conf.routes.default.errorMsgs || {};
        //So many ors because compatibility with version 1
        conf.routes.default = {
            "requestsPerExpiry": conf.routes.default.requestsPerExpiry || conf.options.requestsPerExpiry || 100,
            "completedActionsPerExpiry": conf.routes.default.completedActionsPerExpiry || conf.options.completedActionsPerExpiry || 5,
            "expiry": conf.routes.default.expiry || conf.options.expiry || 900,
            "completedExpiry": conf.routes.default.completedExpiry || conf.routes.default.expiry || conf.options.completedExpiry || conf.options.expiry || 900,
            "errorMsgs": {
                "requests": conf.routes.default.errorMsgs.requests || conf.options.errorMsg || "As you have made too many requests, you are being rate limited.",
                "completed": conf.routes.default.errorMsgs.completed || conf.options.errorMsg || "As you performed too many successful actions, you are being rate limited."
            }
        };

        this.options = {
            //Use provided options or if not, revert to default.
            keyName: conf.options.keyName || "limitrr",
            errorStatusCode: conf.options.errorStatusCode || 429,
            catchErrors: conf.options.catchErrors || true
        };

        this.routes = conf.routes;
        this._setDefaultToUndefined(this.routes)
        this.db = new Redis({
            port: conf.redis.port || 6379,
            host: conf.redis.host || "127.0.0.1",
            family: conf.redis.family || 4,
            password: conf.redis.password || "",
            db: conf.redis.db || 0
        });
        this.db.on("error", (err) => {
            if (this.options.catchErrors) {
                this.isConnected = false;
                console.error("[limitrr] Couldn't connect to redis:", err);
            } else {
                throw new Error(err);
            }
        });
    }

    getRoutes() {
        return this.routes;
    }

    getOptions() {
        return this.options;
    }

    getIp() {
        return (req, res, next) => {
            req.realIp =
                req.headers["cf-connecting-ip"] ||
                req.headers["x-forwarded-for"] ||
                req.connection.remoteAddress;
            next();
        };
    }

    limit(obj = {
        "route": "default"
    }) {
        return (req, res, next) => {
            let route;
            if (typeof obj === "string") {
                route = obj;
            } else {
                route = obj.route;
            }
            let {
                test
            } = obj;
            let ip =
                req.headers["cf-connecting-ip"] ||
                req.headers["x-forwarded-for"] ||
                req.connection.remoteAddress;
            if (test) {
                ip = "test";
            }
            this.db.multi()
                .get(`limitrr:${this.options.keyName}:${ip}:${route}:requests`)
                .get(`limitrr:${this.options.keyName}:${ip}:${route}:completed`)
                .ttl(`limitrr:${this.options.keyName}:${ip}:${route}:requests`)
                .ttl(`limitrr:${this.options.keyName}:${ip}:${route}:completed`)
                .exec()
                .then((result) => {
                    this.routes[route] = this.routes[route] || this.routes.default;
                    if (result[2][1] < 0) {
                        result[2][1] = 0;
                    }
                    if (result[3][1] < 0) {
                        result[3][1] = 0;
                    }
                    res.header("X-RateLimit-Limit", this.routes[route].requestsPerExpiry);
                    res.header("X-RateLimit-Remaining", (this.routes[route].requestsPerExpiry - result[0][1] - 1));
                    res.header("X-RateLimit-Reset", parseInt(result[2][1]));
                    res.header("X-RateLimit-Limit-Actions", this.routes[route].completedActionsPerExpiry);
                    res.header("X-RateLimit-Remaining-Actions", (this.routes[route].completedActionsPerExpiry - result[1][1]));
                    res.header("X-RateLimit-Reset-Actions", parseInt(result[3][1]));
                    if (parseInt(result[0][1]) >= this.routes[route].requestsPerExpiry) {
                        return res.status(this.options.errorStatusCode).json({
                            error: this.routes[route].errorMsgs.requests
                        });
                    } else if (parseInt(result[1][1]) >= this.routes[route].completedActionsPerExpiry) {
                        return res.status(this.options.errorStatusCode).json({
                            error: this.routes[route].errorMsgs.completed
                        });
                    } else if (result[0][1]) {
                        this.db.incr(`limitrr:${this.options.keyName}:${ip}:${route}:requests`).then((result) => {
                            if (result.length > 0) {
                                next();
                            } else {
                                next("[limitrr] Keystore returned invalid data.")
                            }
                        }).catch((err) => {
                            this._handleError(err, res);
                        });
                    } else {
                        this.db
                            .multi()
                            .incr(`limitrr:${this.options.keyName}:${ip}:${route}:requests`)
                            .expire(
                                `limitrr:${this.options.keyName}:${ip}:${route}:requests`, this.routes[route].expiry
                            )
                            .exec()
                            .then((result) => {
                                if (result.length > 0) {
                                    next();
                                } else {
                                    next("[limitrr] Keystore returned invalid data.")
                                }
                            })
                            .catch((err) => {
                                this._handleError(err, res);
                            });
                    }
                })
                .catch((err) => {
                    this._handleError(err, res);
                });
        };
    }

    complete(obj, o_route = "default") {
        return new Promise((resolve, reject) => {
            let discriminator;
            if (typeof obj === "string") {
                discriminator = obj;
            } else {
                discriminator = obj.discriminator;
            }
            let {
                route
            } = obj;
            if (!route) {
                route = o_route;
            }
            this.routes[route] = this.routes[route] || this.routes.default;
            this.db
                .multi()
                .incr(
                    `limitrr:${this.options.keyName}:${discriminator}:${route}:completed`
                )
                .expire(
                    `limitrr:${this.options.keyName}:${discriminator}:${route}:completed`, this.routes[route].completedExpiry
                )
                .exec()
                .then((result) => {
                    if (result.length > 0) {
                        resolve(true);
                    } else {
                        reject("Keystore returned invalid data.");
                    }
                })
                .catch((err) => {
                    reject(err);
                });
        });
    }

    get(obj, o_route = "default", o_type = false) {
        return new Promise((resolve, reject) => {
            let discriminator;
            if (typeof obj === "string") {
                discriminator = obj;
            } else {
                discriminator = obj.discriminator;
            }
            let {
                route,
                type
            } = obj;
            if (!route) {
                route = o_route;
            }
            if (!type) {
                type = o_type;
            }
            if (!type) {
                this.db.multi()
                    .get(`limitrr:${this.options.keyName}:${discriminator}:${route}:requests`)
                    .get(`limitrr:${this.options.keyName}:${discriminator}:${route}:completed`)
                    .exec()
                    .then((result) => {
                        if (result.length > 0) {
                            resolve({
                                "requests": (parseInt(result[0][1]) || 0),
                                "completed": (parseInt(result[1][1]) || 0)
                            });
                        } else {
                            reject("Keystore returned invalid data");
                        }
                    }).catch((err) => {
                        reject(err);
                    });
            } else {
                this.db.get(`limitrr:${this.options.keyName}:${discriminator}:${route}:${type}`).then((result) => {
                    resolve(parseInt(result) || 0);
                }).catch((err) => {
                    reject(err);
                });
            }

        });
    }

    reset(obj, o_type = false, o_route = "default") {
        return new Promise((resolve, reject) => {
            let discriminator;
            if (typeof obj === "string") {
                discriminator = obj;
            } else {
                discriminator = obj.discriminator;
            }
            let {
                type,
                route
            } = obj;
            if (!type) {
                type = o_type;
            }
            if (!route) {
                route = o_route;
            }
            if (!type) {
                this.db.multi()
                    .del(`limitrr:${this.options.keyName}:${discriminator}:${route}:requests`)
                    .del(`limitrr:${this.options.keyName}:${discriminator}:${route}:completed`)
                    .exec().then((result) => {
                        if (result.length > 0) {
                            resolve(true);
                        } else {
                            reject("Keystore returned invalid data");
                        }
                    }).catch((err) => {
                        reject(err);
                    });
            } else {
                this.db.del(`limitrr:${this.options.keyName}:${discriminator}:${route}:${type}`).then((result) => {
                    if (result) {
                        resolve(true);
                    } else {
                        reject("Keystore returned invalid data");
                    }
                }).catch((err) => {
                    reject(err);
                });
            }

        });
    }

    _handleError(err, res = false) {
        if (this.options.catchErrors) {
            console.error("[limitrr] Error:", err);
        } else {
            throw new Error(err);
        }
        return res.status(500).json({
            error: "Internal Error"
        });
    }

    _setDefaultToUndefined(obj) {
        return new Promise((resolve) => {
            Object.keys(obj).forEach((obj2, index) => {
                if (index <= Object.keys(obj).length - 1) {
                    obj2 = obj[obj2];
                    Object.keys(this.routes.default).forEach((key) => {
                        if (!obj2[key]) {
                            obj2[key] = this.routes.default[key];
                        }
                    });
                } else {
                    this.routes = obj;
                    resolve(obj);
                }
            });
        });
    }
}

module.exports = limitrr;